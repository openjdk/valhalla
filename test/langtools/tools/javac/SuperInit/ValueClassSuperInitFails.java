/*
 * @test /nodynamiccopyright/
 * @bug 8324873
 * @summary Permit additional statements before this/super in constructors
 * @compile/fail/ref=ValueClassSuperInitFails.out -XDrawDiagnostics ValueClassSuperInitFails.java
 * @enablePreview
 */

abstract value class AR<V> implements java.io.Serializable {
    public AR(V initialValue) {
    }

    public AR() {
    }
}

value class ValueClassSuperInitFails extends AR <Object> implements Iterable<Object> {

    private int x;

/// GOOD EXAMPLES

    public ValueClassSuperInitFails() {           // this should be OK
        // super()
    }

    public ValueClassSuperInitFails(Object x) {
        this.x = x.hashCode();          // this should be OK
        // super();  the compiler will introduce the super call at this location
    }

    public ValueClassSuperInitFails(byte x) {
        super();                        // this should be OK
    }

    public ValueClassSuperInitFails(char x) {
        this((int)x);                   // this should be OK
    }

/// FAIL EXAMPLES

    {
        this(1);                        // this should FAIL
    }

    {
        super();                        // this should FAIL
    }

    void normalMethod1() {
        super();                        // this should FAIL
    }

    void normalMethod2() {
        this();                         // this should FAIL
    }

    void normalMethod3() {
        Runnable r = () -> super();     // this should FAIL
    }

    void normalMethod4() {
        Runnable r = () -> this();      // this should FAIL
    }

    public ValueClassSuperInitFails(short x) {
        hashCode();                     // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(float x) {
        this.hashCode();                // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(int x) {
        super.hashCode();               // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(long x) {
        ValueClassSuperInitFails.this.hashCode();      // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(double x) {
        ValueClassSuperInitFails.super.hashCode();     // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(byte[] x) {
        {
            super();                    // this should FAIL
        }
    }

    public ValueClassSuperInitFails(char[] x) {
        if (x.length == 0)
            return;                     // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(short[] x) {
        this.x = x.length;              // this should be OK
        //super();
    }

    public ValueClassSuperInitFails(float[] x) {
        System.identityHashCode(this);  // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(int[] x) {
        this(this);                     // this should FAIL
    }

    public ValueClassSuperInitFails(long[] x) {
        this(Object.this);              // this should FAIL
    }

    public ValueClassSuperInitFails(double[] x) {
        Iterable.super.spliterator();   // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(byte[][] x) {
        super(new Object() {
            {
                super();                // this should FAIL
            }
        });
    }

    public ValueClassSuperInitFails(char[][] x) {
        new Inner1();                   // this should FAIL
        //super();
    }

    class Inner1 {
    }

    record Record1(int value) {
        Record1(float x) {              // this should FAIL
        }
    }

    record Record2(int value) {
        Record2(float x) {              // this should FAIL
            super();
        }
    }

    @Override
    public java.util.Iterator<Object> iterator() {
        return null;
    }

    public ValueClassSuperInitFails(short[][] x) {
        class Foo {
            Foo() {
                ValueClassSuperInitFails.this.hashCode();
            }
        };
        new Foo();                      // this should FAIL
        //super();
    }

    public ValueClassSuperInitFails(float[][] x) {
        Runnable r = () -> {
            super();                    // this should FAIL
        };
    }

    public ValueClassSuperInitFails(int[][] z) {
        super((Runnable)() -> x);       // this should FAIL
    }

    public ValueClassSuperInitFails(long[][] z) {
        super(new Inner1());            // this should FAIL
    }
}
