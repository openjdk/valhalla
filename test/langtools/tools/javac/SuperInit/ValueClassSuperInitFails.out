ValueClassSuperInitFails.java:67:9: compiler.err.cant.ref.before.ctor.called: hashCode()
ValueClassSuperInitFails.java:72:9: compiler.err.cant.ref.before.ctor.called: this
ValueClassSuperInitFails.java:77:9: compiler.err.cant.ref.before.ctor.called: super
ValueClassSuperInitFails.java:82:33: compiler.err.cant.ref.before.ctor.called: this
ValueClassSuperInitFails.java:87:33: compiler.err.cant.ref.before.ctor.called: super
ValueClassSuperInitFails.java:109:33: compiler.err.cant.ref.before.ctor.called: this
ValueClassSuperInitFails.java:114:14: compiler.err.cant.ref.before.ctor.called: this
ValueClassSuperInitFails.java:118:20: compiler.err.not.encl.class: java.lang.Object
ValueClassSuperInitFails.java:122:17: compiler.err.cant.ref.before.ctor.called: super
ValueClassSuperInitFails.java:129:22: compiler.err.call.must.only.appear.in.ctor
ValueClassSuperInitFails.java:135:9: compiler.err.cant.ref.before.ctor.called: this
ValueClassSuperInitFails.java:143:9: compiler.err.non.canonical.constructor.invoke.another.constructor: ValueClassSuperInitFails.Record1
ValueClassSuperInitFails.java:148:9: compiler.err.non.canonical.constructor.invoke.another.constructor: ValueClassSuperInitFails.Record2
ValueClassSuperInitFails.java:164:9: compiler.err.cant.ref.before.ctor.called: this
ValueClassSuperInitFails.java:175:31: compiler.err.cant.ref.before.ctor.called: x
ValueClassSuperInitFails.java:179:15: compiler.err.cant.ref.before.ctor.called: this
ValueClassSuperInitFails.java:43:13: compiler.err.call.must.only.appear.in.ctor
ValueClassSuperInitFails.java:47:14: compiler.err.call.must.only.appear.in.ctor
ValueClassSuperInitFails.java:51:14: compiler.err.call.must.only.appear.in.ctor
ValueClassSuperInitFails.java:55:13: compiler.err.call.must.only.appear.in.ctor
ValueClassSuperInitFails.java:59:33: compiler.err.call.must.only.appear.in.ctor
ValueClassSuperInitFails.java:63:32: compiler.err.call.must.only.appear.in.ctor
ValueClassSuperInitFails.java:93:18: compiler.err.ctor.calls.not.allowed.here
ValueClassSuperInitFails.java:99:13: compiler.err.return.before.superclass.initialized
ValueClassSuperInitFails.java:170:18: compiler.err.ctor.calls.not.allowed.here
- compiler.note.preview.filename: ValueClassSuperInitFails.java, DEFAULT
- compiler.note.preview.recompile
25 errors
